### ISSUE TITLE
Questions regarding Quantization Aware Training

### ISSUE THREAD

**Issue by tabeareichmann** [tensorflow/model-optimization#436]
Hello,
we are looking into using quantization aware training for a research project to determine the impact of quantization during training on convergence rates and runtimes. We are though not yet fully convinced that this is the right tool. Could you please clarify the following points:
.) If a layer is quantized during quantization aware training, this means inputs and weights are quantized and all operations including activation function are quantized and then, before returning, the outputs are de-quantized to a precision compatible with the next layer. Is this understanding correct?
.) Is it compatible with the Tensorboard profiler?
.) Does quantization aware training, in principle, lead to a speedup during training in your general experience or is this impossible due to it being solely a simulation?
.) Can you point us to a resource on how to add custom quantizers and datatypes to tensorflow s.t. they are GPU compatible?
Thank you very much!

**Comment by nutsiepully** [comment]
Hi @tabeareichmann,

Thanks for the great questions. Please find answers below.

1) The quantization and dequantization operation is for a specific tensor, but placement of the operation takes the layer into account. The second half of [this](https://www.youtube.com/watch?v=2tnWHCxP8Bk) talk explains this more. There's another detailed talk on the way, though this has all the relevant content.
   Basically, the inputs and weights are quantized/dequantized. This leads to the tensors falling at the boundaries of int8 computations even though they are themselves float32. This leads to almost exact emulation of int8 arithmetic during training. The placement of the quant operations takes into account the activations, and layers etc. to ensure they match inference execution.

2) Yes, it just outputs a TF keras model. So you can train and profile it the same on TensorBoard.

3) No, it actually slows training down marginally. The goal is to speed up inference. So once you train with QAT, you can create an inference model which is quantized and highly performant while retaining accuracy. The aforementioned talk, and this [blog](https://blog.tensorflow.org/2020/04/quantization-aware-training-with-tensorflow-model-optimization-toolkit.html) should help explain this further.

4) [This](https://www.tensorflow.org/model_optimization/guide/quantization/training_comprehensive_guide#experiment_with_quantization) section should help explain it. You would need to use your GPU specific `Quantizer`s and `QuantizeConfig`s. Also, read the existing code in the `default_8bit` directory. Something similar to that for GPUs would need to be written.

You're welcome :)

Closing this for now. Feel free to re-open if you have more questions.
